docs/source/dev/functions.rst

```diff
@@ -0,0 +1,475 @@
+Argument and Keyword Naming Conventions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If an argument or keyword is intended to be an analog for an argument to
+a Windows API call then it should follow the same naming convention as
+the documented function does. The `WaitForSingleObject` function for example
+takes two arguments according to the MSDN documentation which when translated
+to Python would look like this:
+
+.. code-block:: python
+
+   def WaitForSingleObject(hHandle, dwMilliseconds):
+       pass
+
+Any argument or keyword which is not directly related to an input to a Windows
+API should instead use the standard PEP8 naming conventions:
+
+.. code-block:: python
+
+   def WaitForSingleObject(hHandle, dwMilliseconds, other_keyword=None):
+       pass
+
+
+Internal Variables
+~~~~~~~~~~~~~~~~~~
+
+Like arguments or keywords variables should be named either using `camelCase`
+if they're intended to map to a value passed into a Windows API call or using
+the `name_with_underscores` convention in other cases.  Here's an example of
+the two:
+
+.. code-block:: python
+
+   def UnlockFileEx(...):
+
+        # internal variables
+        ffi, library = dist.load()
+
+        # lpOverlapped is a Windows structure
+        if lpOverlapped is None:
+            lpOverlapped = ffi.new("OVERLAPPED[]", [{"hEvent": hFile}])
+
+
+Documentation
+-------------
+
+This section covers the basics of documenting functions in pywincffi.  The
+below mostly applies to how Windows functions should be documented but should
+generally apply elsewhere in the project too.
+
+Basic Layout
+~~~~~~~~~~~~
+
+The layout of the documentation string for each function should be consistent
+throughout the project.  This generally makes it easier to understand but also
+harder to miss more critical information.  Below is an annotated example
+of a fake Windows function:
+
+.. code-block:: python
+
+   def AWindowsFunction(...):
+       """
+       First few sentences should tell someone what AWindowsFunction
+       does.  This can usually be pulled from the MSDN documentation but
+       is usually shorter and more concise.
+
+       .. seealso::
+
+          <url pointing to the msdn reference for AWindowsFunction>
+          <url pointing to a use case or other usefull information>
+
+       :param <python type> variable_name:
+           Some information about what variable_name is.  Again, can be pulled
+           from the msdn documentation but should be concise as someone can
+           always go read the msdn documentation.  This information should
+           always state key differences, if there are any, between what
+           the C api call normally expects and what the wrapper does.
+
+       <additional keyword or argument documentation>
+
+       :raises SomeException:
+           Information about under what condition(s) SomeException may be
+           raised.  SomeException should be something that's raised directly
+           by AWindowsFunction.
+
```

Comment:
I saw a comment on existing code, which I used myself, describing `:rtype:` as well. Should it be here, if if optional?